Make `BorrowedBuf` and `BorrowedCursor` generic over the data

joshtriplett opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/149749)*
<!-- homu-ignore:end -->

Replace uses of `u8` with a generic type. This will allow reusing
`BorrowedBuf` elsewhere in the standard library, for purposes other than
byte buffers.

This currently requires `T: Copy`, to prevent `T: Drop`, which would require additional careful handling. We can consider relaxing that restriction in the future if we make `T: Drop` work (without pessimizing `Copy` types like `u8`).
    
`ensure_init` currently continues to require `u8`. We could potentially generalize it to `T: Default`, but we'd need to ensure it produced the same or better assembly for `u8`.


-----

As discussed in libs-api; cc @Amanieu.

joshtriplett committed:
Make `BorrowedBuf` and `BorrowedCursor` generic over the data

Replace uses of `u8` with a generic type. This will allow reusing
`BorrowedBuf` elsewhere in the standard library, for purposes other than
byte buffers.

This currently requires `T: Copy`, to prevent `T: Drop`, which would
require additional careful handling. We can consider relaxing that
restriction in the future if we make `T: Drop` work (without pessimizing
`Copy` types like `u8`).

`ensure_init` currently continues to require `u8`. We could potentially
generalize it to `T: Default`, but we'd need to ensure it produced the
same or better assembly for `u8`.

Amanieu reviewed (commented):
r=me for the rest

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149749/ba2142a19c1acabe770433418fa395b8a0ef47a4..c1509316be4000e53435720424aaba5058586838)*

scottmcm reviewed (approved):
I think it'd be fine to land this PR for generic-izing the type, then we can land another PR right afterwards to remove the default and update the uses.

Code looks good to me, and I tried to scan for anything making unsaid assumptions about `T` (being Copy, Freeze, etc) and didn't find any that weren't already updated.

So `r=me` with conflicts fixed.

---

Aside: excited to try this out with generic types.  Might be a great way to solve some of the `Iterator::next_chunk` or `Vec::push_within_existing_capacity` scenarios.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149749/ba2142a19c1acabe770433418fa395b8a0ef47a4..c1509316be4000e53435720424aaba5058586838)*

briansmith commented on library/core/src/io/borrowed_buf.rs:
```diff
@@ -24,14 +24,16 @@ use crate::mem::{self, MaybeUninit};
 /// The lifetime `'data` is a bound on the lifetime of the underlying data.
-pub struct BorrowedBuf<'data> {
+///
+/// The type defaults to managing bytes, but can manage any type of data.
```
*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149749/ba2142a19c1acabe770433418fa395b8a0ef47a4..c1509316be4000e53435720424aaba5058586838)*

It seems prudent to have tests with a `T` that isn't the same size as `u8` to ensure that there are no byte-size vs. slice-length issues.

Similarly, it would be nice to have some test for non-`Copy` `T`. Since the `BorrowedBuf` doesn't own any `T` then no special drop handling is necessary. Is it worth calling this out? I seem to remember in the discussion of doing this that people thought that this could be done for only `T: Copy` presumably because of drop checking.

briansmith replied:
I think `element` is better than `data` here.

briansmith replied:
In particular, it would be good to clarify (improve documentation and add a test) of `BorrowedBuf::clear()` with a non-empty `BorrowedBuf` of `T: Drop`. It seems wrong to "forget" that some values that need to be dropped are there, as then a new BorrowedCursor would overwrite them without dropping them, right?

briansmith replied:
This was all resolved except for the part about having tests for types that have a size other than 1. I think in particular having tests for ZSTs would be useful.

joshtriplett replied:
Added tests for `u64` and `()`.

briansmith commented on library/core/src/io/borrowed_buf.rs:
```diff
@@ -3,11 +3,11 @@
 /// This type makes it safer to work with `MaybeUninit` buffers, such as to read into a buffer
-/// without having to initialize it first. It tracks the region of bytes that have been filled and
-/// the region that remains uninitialized.
+/// without having to initialize it first. It tracks the region of data that has been filled and the
```
*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149749/ba2142a19c1acabe770433418fa395b8a0ef47a4..c1509316be4000e53435720424aaba5058586838)*

I think "elements that have" is better (clearer) than "data that has" here.

briansmith commented on library/core/src/io/borrowed_buf.rs:
```diff
@@ -210,22 +214,22 @@ impl<'a> BorrowedCursor<'a> {
     /// # Safety
     ///
-    /// The caller must not uninitialize any previously initialized bytes.
+    /// The caller must not uninitialize any previously initialized data.
```
*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149749/ba2142a19c1acabe770433418fa395b8a0ef47a4..c1509316be4000e53435720424aaba5058586838)*

elements?

briansmith replied:
Consider the case where `T: Drop`. Then `as_mut` now has an additional hazard: For any initialized element, overwriting the element must drop the old one, right?

briansmith commented on library/core/src/io/borrowed_buf.rs:
```diff
@@ -239,7 +243,10 @@ impl<'a> BorrowedCursor<'a> {
     /// Panics if `self.capacity()` is less than `buf.len()`.
     #[inline]
-    pub fn append(&mut self, buf: &[u8]) {
+    pub fn append(&mut self, buf: &[T])
```
*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149749/ba2142a19c1acabe770433418fa395b8a0ef47a4..c1509316be4000e53435720424aaba5058586838)*

Perhaps this should be renamed `append_copy_of_slice`, analogous to `slice::write_copy_of_slice`, to make room for a future method that doesn't require `T: Copy`?

joshtriplett replied:
@briansmith Are you thinking of a method that would work with `T: Clone` instead? Or, one that takes ownership and does a move? I think both of those would be the methods that should get more elaborate names; `append` seems like the most natural operation here.

briansmith replied:
> @briansmith Are you thinking of a method that would work with `T: Clone` instead? Or, one that takes ownership and does a move? I think both of those would be the methods that should get more elaborate names; `append` seems like the most natural operation here.

It matters less now that we have `T: Copy` constraints.

briansmith commented on library/core/src/io/borrowed_buf.rs:
```diff
@@ -41,13 +43,13 @@ impl Debug for BorrowedBuf<'_> {
 
 /// Creates a new `BorrowedBuf` from a fully initialized slice.
-impl<'data> From<&'data mut [u8]> for BorrowedBuf<'data> {
+impl<'data, T> From<&'data mut [T]> for BorrowedBuf<'data, T> {
```
*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149749/ba2142a19c1acabe770433418fa395b8a0ef47a4..c1509316be4000e53435720424aaba5058586838)*

If `T: Drop` then this is pretty dangerous, because it immediately "forgets" that all the elements need to be dropped (before being overwritten).

briansmith commented on library/core/src/io/borrowed_buf.rs:
```diff
@@ -23,16 +23,18 @@ use crate::ptr;
 }
 
-impl Debug for BorrowedBuf<'_> {
+impl<T: Copy> Debug for BorrowedBuf<'_, T> {
```
I think this is better without `T: Copy`; otherwise, things that contain a `BorrowedBuf` would not be able to derive `Debug` if they are also generic over `T`, without having the same constraint.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149749/783eb8c8682ddde0807c60ed8293670ef523794f..f7b1581187a82c87f53e701d5dc9d904df62915c)*

joshtriplett replied:
Fixed (for both types).

briansmith commented on library/core/src/io/borrowed_buf.rs:
```diff
@@ -180,35 +184,40 @@ impl<'data> BorrowedBuf<'data> {
 }
 
-impl<'a> BorrowedCursor<'a> {
+impl<T: Copy> Debug for BorrowedCursor<'_, T> {
```
Ditto.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149749/783eb8c8682ddde0807c60ed8293670ef523794f..f7b1581187a82c87f53e701d5dc9d904df62915c)*

joshtriplett committed:
Avoid requiring `T: Copy` for `Debug` impls

joshtriplett committed:
BorrowedBuf: Rewrite docs to say "elements" rather than "data" or "bytes"

joshtriplett committed:
Remove default `T = u8` from `BorrowedBuf` and `BorrowedCursor`

This default simplified the commit making these types generic, but we
don't actually want the default. Rewrite every user to specify `u8`
explicitly.

joshtriplett commented on library/core/src/io/borrowed_buf.rs:
```diff
@@ -24,14 +24,16 @@ use crate::mem::{self, MaybeUninit};
-pub struct BorrowedBuf<'data> {
+///
+/// The type defaults to managing bytes, but can manage any type of data.
+pub struct BorrowedBuf<'data, T = u8> {
```
I've added a commit to remove the default.

joshtriplett commented on library/core/src/io/borrowed_buf.rs:
```diff
@@ -24,14 +24,16 @@ use crate::mem::{self, MaybeUninit};
 }
 
-impl Debug for BorrowedBuf<'_> {
+impl<T> Debug for BorrowedBuf<'_, T> {
```
I don't think we will generally want the borrowed buffer attempting to print its contents.

If we want to change that, then yes, that wouldn't be in this PR.

joshtriplett committed:
BorrowedBuf: Add tests for non-u8 buffers/cursors

joshtriplett committed:
BorrowedBuf: Add tests for ZST buffers/cursors

joshtriplett commented:
@rustbot ready

This is ready for review.

I would suggest looking at it commit by commit, as the change to remove the default `T = u8` is *very* large and noisy, but largely mechanical.

joshtriplett commented:
r? libs

joshtriplett committed:
Remove redundant `T: Copy` bound in `BorrowedCursor::append`

Now that the `impl` block has a bound, the one on `append` became
redundant.

joshtriplett commented:
Pushing one additional commit with a minor fix (no semantic effect).

Caught by an experiment trying PR review with an AI; used for review only, no code. It noted seven issues, of which 1 was real (this one), 1 was obviously wrong (suggesting a test that would have called `set_init` with the buffer still uninitialized), 1 was *subtly* wrong (it would have provided a safe [`conjure_zst`](https://doc.rust-lang.org/std/mem/fn.conjure_zst.html)), and 4 were quibbles about code refactoring in diff context that this PR doesn't change.

joboet commented:
Looks fine!
@bors r+ rollup
